airbyte-integrations/connectors/source-mysql/src/main/java/io/airbyte/integrations/source/mysql/initialsync/MySqlInitialLoadStreamStateManager.java

```diff
@@ -0,0 +1,113 @@
+import io.airbyte.commons.json.Jsons;
+import io.airbyte.integrations.source.mysql.initialsync.MySqlInitialReadUtil.InitialLoadStreams;
+import io.airbyte.integrations.source.mysql.initialsync.MySqlInitialReadUtil.PrimaryKeyInfo;
+import io.airbyte.integrations.source.mysql.internal.models.PrimaryKeyLoadStatus;
+import io.airbyte.integrations.source.relationaldb.models.DbStreamState;
+import io.airbyte.protocol.models.v0.AirbyteStateMessage;
+import io.airbyte.protocol.models.v0.AirbyteStateMessage.AirbyteStateType;
+import io.airbyte.protocol.models.v0.AirbyteStreamState;
+import io.airbyte.protocol.models.v0.ConfiguredAirbyteCatalog;
+import io.airbyte.protocol.models.v0.StreamDescriptor;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This state manager extends the StreamStateManager to enable writing the state_type and version
+ * keys to the stream state when they're going through the iterator Once we have verified that
+ * expanding StreamStateManager itself to include this functionality, this class will be removed
+ */
+public class MySqlInitialLoadStreamStateManager implements MySqlInitialLoadStateManager {
+
+  private final Map<io.airbyte.protocol.models.AirbyteStreamNameNamespacePair, PrimaryKeyLoadStatus> pairToPrimaryKeyLoadStatus;
+
+  // Map of pair to the primary key info (field name & data type) associated with it.
+  private final Map<io.airbyte.protocol.models.AirbyteStreamNameNamespacePair, PrimaryKeyInfo> pairToPrimaryKeyInfo;
+
+  private final Set<io.airbyte.protocol.models.AirbyteStreamNameNamespacePair> streamsThatHaveCompletedSnapshot;
+  private static final Logger LOGGER = LoggerFactory.getLogger(MySqlInitialLoadStreamStateManager.class);
+
+  public MySqlInitialLoadStreamStateManager(final ConfiguredAirbyteCatalog catalog,
+                                            final InitialLoadStreams initialLoadStreams,
+                                            final Map<io.airbyte.protocol.models.AirbyteStreamNameNamespacePair, PrimaryKeyInfo> pairToPrimaryKeyInfo) {
+    this.pairToPrimaryKeyInfo = pairToPrimaryKeyInfo;
+    this.pairToPrimaryKeyLoadStatus = initPairToPrimaryKeyLoadStatusMap(initialLoadStreams.pairToInitialLoadStatus());
+    this.streamsThatHaveCompletedSnapshot = initStreamsCompletedSnapshot(initialLoadStreams, catalog);
+  }
+
+  @Override
+  public void updatePrimaryKeyLoadState(final io.airbyte.protocol.models.AirbyteStreamNameNamespacePair pair,
+                                        final PrimaryKeyLoadStatus pkLoadStatus) {
+    pairToPrimaryKeyLoadStatus.put(pair, pkLoadStatus);
+  }
+
+  public AirbyteStateMessage createFinalStateMessage(final io.airbyte.protocol.models.AirbyteStreamNameNamespacePair pair,
+                                                     final JsonNode streamStateForIncrementalRun) {
+    streamsThatHaveCompletedSnapshot.add(pair);
```

Comment:
this is not required for stream state manager. It is only required for global state manager